// @dcl/catalyst-contracts/dist/ListAbi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listAbiItems = exports.denylistNamesDeployments = exports.poiDeployments = void 0;
exports.poiDeployments = {
    '80001' /* mumbai  */: '0x08E5a5288D6bBa9383724C57175C03A37fe83A2A',
    '137' /* polygon */: '0xFEC09d5C192aaf7Ec7E2C89Cc8D3224138391B2E'
};
exports.denylistNamesDeployments = {
    '5' /* goerli  */: '0x71c84760df0537f7db286274817462dc2e6c1366',
    '1' /* mainnet */: '0x0c4c90a4f29872a2e9ef4c4be3d419792bca9a36'
};
exports.listAbiItems = [
    {
        constant: true,
        inputs: [],
        name: 'ADD_ROLE',
        outputs: [
            {
                name: '',
                type: 'bytes32'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'name',
        outputs: [
            {
                name: '',
                type: 'string'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'hasInitialized',
        outputs: [
            {
                name: '',
                type: 'bool'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [
            {
                name: '_script',
                type: 'bytes'
            }
        ],
        name: 'getEVMScriptExecutor',
        outputs: [
            {
                name: '',
                type: 'address'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'getRecoveryVault',
        outputs: [
            {
                name: '',
                type: 'address'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'REMOVE_ROLE',
        outputs: [
            {
                name: '',
                type: 'bytes32'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [
            {
                name: '',
                type: 'uint256'
            }
        ],
        name: 'values',
        outputs: [
            {
                name: '',
                type: 'string'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [
            {
                name: 'token',
                type: 'address'
            }
        ],
        name: 'allowRecoverability',
        outputs: [
            {
                name: '',
                type: 'bool'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: false,
        inputs: [
            {
                name: '_value',
                type: 'string'
            }
        ],
        name: 'remove',
        outputs: [],
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'appId',
        outputs: [
            {
                name: '',
                type: 'bytes32'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'getInitializationBlock',
        outputs: [
            {
                name: '',
                type: 'uint256'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'getTypeHash',
        outputs: [
            {
                name: '',
                type: 'bytes32'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'listType',
        outputs: [
            {
                name: '',
                type: 'string'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'size',
        outputs: [
            {
                name: '',
                type: 'uint256'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [
            {
                name: '_index',
                type: 'uint256'
            }
        ],
        name: 'get',
        outputs: [
            {
                name: '',
                type: 'string'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'symbol',
        outputs: [
            {
                name: '',
                type: 'string'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: false,
        inputs: [
            {
                name: '_token',
                type: 'address'
            }
        ],
        name: 'transferToVault',
        outputs: [],
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function'
    },
    {
        constant: true,
        inputs: [
            {
                name: '_sender',
                type: 'address'
            },
            {
                name: '_role',
                type: 'bytes32'
            },
            {
                name: '_params',
                type: 'uint256[]'
            }
        ],
        name: 'canPerform',
        outputs: [
            {
                name: '',
                type: 'bool'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'getEVMScriptRegistry',
        outputs: [
            {
                name: '',
                type: 'address'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: false,
        inputs: [
            {
                name: '_name',
                type: 'string'
            },
            {
                name: '_symbol',
                type: 'string'
            },
            {
                name: '_type',
                type: 'string'
            }
        ],
        name: 'initialize',
        outputs: [],
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function'
    },
    {
        constant: false,
        inputs: [
            {
                name: '_value',
                type: 'string'
            }
        ],
        name: 'add',
        outputs: [],
        payable: false,
        stateMutability: 'nonpayable',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'kernel',
        outputs: [
            {
                name: '',
                type: 'address'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'isPetrified',
        outputs: [
            {
                name: '',
                type: 'bool'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                name: '_caller',
                type: 'address'
            },
            {
                indexed: false,
                name: '_value',
                type: 'string'
            }
        ],
        name: 'Add',
        type: 'event'
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                name: '_caller',
                type: 'address'
            },
            {
                indexed: false,
                name: '_value',
                type: 'string'
            }
        ],
        name: 'Remove',
        type: 'event'
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                name: 'executor',
                type: 'address'
            },
            {
                indexed: false,
                name: 'script',
                type: 'bytes'
            },
            {
                indexed: false,
                name: 'input',
                type: 'bytes'
            },
            {
                indexed: false,
                name: 'returnData',
                type: 'bytes'
            }
        ],
        name: 'ScriptResult',
        type: 'event'
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                name: 'vault',
                type: 'address'
            },
            {
                indexed: true,
                name: 'token',
                type: 'address'
            },
            {
                indexed: false,
                name: 'amount',
                type: 'uint256'
            }
        ],
        name: 'RecoverToVault',
        type: 'event'
    }
];
