// @dcl/sdk/message-bus.js
import { onCommsMessage } from './observables';
import * as communicationsController from '~system/CommunicationsController';
export class MessageBus {
    constructor() {
        this.messageQueue = [];
        this.flushing = false;
    }
    on(message, callback) {
        return onCommsMessage.add((e) => {
            const m = JSON.parse(e.message);
            if (m.message === message) {
                callback(m.payload, e.sender);
            }
        });
    }
    sendRaw(message) {
        this.messageQueue.push(message);
        this.flush();
    }
    emit(message, payload) {
        const messageToSend = JSON.stringify({ message, payload });
        this.sendRaw(messageToSend);
        onCommsMessage.notifyObservers({ message: messageToSend, sender: 'self' });
    }
    flush() {
        if (this.messageQueue.length === 0)
            return;
        if (this.flushing)
            return;
        const message = this.messageQueue.shift();
        this.flushing = true;
        communicationsController.send({ message }).then((_) => {
            this.flushing = false;
            this.flush();
        }, (_) => {
            this.flushing = false;
        });
    }
}
