// @dcl/sdk/observables.js
import { Observable } from './internal/Observable';
import { subscribe } from '~system/EngineApi';
let subscribeFunction = subscribe;
function createSubscriber(eventName) {
    return () => {
        subscribeFunction({ eventId: eventName }).catch(console.error);
    };
}
export const onEnterSceneObservable = new Observable(createSubscriber('onEnterScene'));
export const onEnterScene = onEnterSceneObservable;
export const onLeaveSceneObservable = new Observable(createSubscriber('onLeaveScene'));
export const onLeaveScene = onLeaveSceneObservable;
export const onSceneReadyObservable = new Observable(createSubscriber('sceneStart'));
export const onPlayerExpressionObservable = new Observable(createSubscriber('playerExpression'));
export const onVideoEvent = new Observable(createSubscriber('videoEvent'));
export const onProfileChanged = new Observable(createSubscriber('profileChanged'));
export const onPlayerConnectedObservable = new Observable(createSubscriber('playerConnected'));
export const onPlayerDisconnectedObservable = new Observable(createSubscriber('playerDisconnected'));
export const onRealmChangedObservable = new Observable(createSubscriber('onRealmChanged'));
export const onPlayerClickedObservable = new Observable(createSubscriber('playerClicked'));
export const onCommsMessage = new Observable(createSubscriber('comms'));
export function setSubscribeFunction(fn) {
    subscribeFunction = fn;
}
export async function pollEvents(sendBatch) {
    const { events } = await sendBatch({ actions: [] });
    for (const e of events) {
        if (e.generic) {
            const data = JSON.parse(e.generic.eventData);
            switch (e.generic.eventId) {
                case 'onEnterScene': {
                    onEnterSceneObservable.notifyObservers(data);
                    break;
                }
                case 'onLeaveScene': {
                    onLeaveSceneObservable.notifyObservers(data);
                    break;
                }
                case 'sceneStart': {
                    onSceneReadyObservable.notifyObservers(data);
                    break;
                }
                case 'playerExpression': {
                    onPlayerExpressionObservable.notifyObservers(data);
                    break;
                }
                case 'videoEvent': {
                    const videoData = data;
                    onVideoEvent.notifyObservers(videoData);
                    break;
                }
                case 'profileChanged': {
                    onProfileChanged.notifyObservers(data);
                    break;
                }
                case 'playerConnected': {
                    onPlayerConnectedObservable.notifyObservers(data);
                    break;
                }
                case 'playerDisconnected': {
                    onPlayerDisconnectedObservable.notifyObservers(data);
                    break;
                }
                case 'onRealmChanged': {
                    onRealmChangedObservable.notifyObservers(data);
                    break;
                }
                case 'playerClicked': {
                    onPlayerClickedObservable.notifyObservers(data);
                    break;
                }
                case 'comms': {
                    onCommsMessage.notifyObservers(data);
                    break;
                }
            }
        }
    }
}
