// @dcl/protocol/out-js/decentraland/kernel/apis/communications_controller.gen.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommunicationsControllerServiceDefinition = exports.RealSendResponse = exports.RealSendRequest = exports.protobufPackage = void 0;
/* eslint-disable */
const minimal_1 = __importDefault(require("protobufjs/minimal"));
exports.protobufPackage = "decentraland.kernel.apis";
function createBaseRealSendRequest() {
    return { message: "" };
}
var RealSendRequest;
(function (RealSendRequest) {
    function encode(message, writer = minimal_1.default.Writer.create()) {
        if (message.message !== "") {
            writer.uint32(10).string(message.message);
        }
        return writer;
    }
    RealSendRequest.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRealSendRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    if (tag !== 10) {
                        break;
                    }
                    message.message = reader.string();
                    continue;
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    RealSendRequest.decode = decode;
    function fromJSON(object) {
        return { message: isSet(object.message) ? String(object.message) : "" };
    }
    RealSendRequest.fromJSON = fromJSON;
    function toJSON(message) {
        const obj = {};
        message.message !== undefined && (obj.message = message.message);
        return obj;
    }
    RealSendRequest.toJSON = toJSON;
    function create(base) {
        return RealSendRequest.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    RealSendRequest.create = create;
    function fromPartial(object) {
        var _a;
        const message = createBaseRealSendRequest();
        message.message = (_a = object.message) !== null && _a !== void 0 ? _a : "";
        return message;
    }
    RealSendRequest.fromPartial = fromPartial;
})(RealSendRequest = exports.RealSendRequest || (exports.RealSendRequest = {}));
function createBaseRealSendResponse() {
    return {};
}
var RealSendResponse;
(function (RealSendResponse) {
    function encode(_, writer = minimal_1.default.Writer.create()) {
        return writer;
    }
    RealSendResponse.encode = encode;
    function decode(input, length) {
        const reader = input instanceof minimal_1.default.Reader ? input : minimal_1.default.Reader.create(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRealSendResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skipType(tag & 7);
        }
        return message;
    }
    RealSendResponse.decode = decode;
    function fromJSON(_) {
        return {};
    }
    RealSendResponse.fromJSON = fromJSON;
    function toJSON(_) {
        const obj = {};
        return obj;
    }
    RealSendResponse.toJSON = toJSON;
    function create(base) {
        return RealSendResponse.fromPartial(base !== null && base !== void 0 ? base : {});
    }
    RealSendResponse.create = create;
    function fromPartial(_) {
        const message = createBaseRealSendResponse();
        return message;
    }
    RealSendResponse.fromPartial = fromPartial;
})(RealSendResponse = exports.RealSendResponse || (exports.RealSendResponse = {}));
exports.CommunicationsControllerServiceDefinition = {
    name: "CommunicationsControllerService",
    fullName: "decentraland.kernel.apis.CommunicationsControllerService",
    methods: {
        /**
         * @deprecated - This API should use a bidirectional binary stream in sdk7
         * https://github.com/decentraland/sdk/issues/582
         */
        send: {
            name: "Send",
            requestType: RealSendRequest,
            requestStream: false,
            responseType: RealSendResponse,
            responseStream: false,
            options: {},
        },
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
