// @dcl/protocol/out-ts/decentraland/kernel/apis/runtime.gen.ts
/* eslint-disable */
import _m0 from "protobufjs/minimal";
import { ContentMapping } from "../../common/content_mapping.gen";

export const protobufPackage = "decentraland.kernel.apis";

export interface RealmInfo {
  baseUrl: string;
  realmName: string;
  networkId: number;
  commsAdapter: string;
  isPreview: boolean;
}

export interface GetRealmResponse {
  realmInfo?: RealmInfo | undefined;
}

export interface GetWorldTimeResponse {
  seconds: number;
}

export interface GetRealmRequest {
}

export interface GetWorldTimeRequest {
}

export interface ReadFileRequest {
  /** name of the deployed file */
  fileName: string;
}

export interface ReadFileResponse {
  /** contents of the file */
  content: Uint8Array;
  /** deployed hash/CID */
  hash: string;
}

export interface CurrentSceneEntityRequest {
}

export interface CurrentSceneEntityResponse {
  /** this is either the entityId or the full URN of the scene that is running */
  urn: string;
  /** contents of the deployed entities */
  content: ContentMapping[];
  /** JSON serialization of the entity.metadata field */
  metadataJson: string;
  /** baseUrl used to resolve all content files */
  baseUrl: string;
}

function createBaseRealmInfo(): RealmInfo {
  return { baseUrl: "", realmName: "", networkId: 0, commsAdapter: "", isPreview: false };
}

export namespace RealmInfo {
  export function encode(message: RealmInfo, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.baseUrl !== "") {
      writer.uint32(10).string(message.baseUrl);
    }
    if (message.realmName !== "") {
      writer.uint32(18).string(message.realmName);
    }
    if (message.networkId !== 0) {
      writer.uint32(24).int32(message.networkId);
    }
    if (message.commsAdapter !== "") {
      writer.uint32(34).string(message.commsAdapter);
    }
    if (message.isPreview === true) {
      writer.uint32(40).bool(message.isPreview);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): RealmInfo {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseRealmInfo();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.baseUrl = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.realmName = reader.string();
          continue;
        case 3:
          if (tag !== 24) {
            break;
          }

          message.networkId = reader.int32();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.commsAdapter = reader.string();
          continue;
        case 5:
          if (tag !== 40) {
            break;
          }

          message.isPreview = reader.bool();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): RealmInfo {
    return {
      baseUrl: isSet(object.baseUrl) ? String(object.baseUrl) : "",
      realmName: isSet(object.realmName) ? String(object.realmName) : "",
      networkId: isSet(object.networkId) ? Number(object.networkId) : 0,
      commsAdapter: isSet(object.commsAdapter) ? String(object.commsAdapter) : "",
      isPreview: isSet(object.isPreview) ? Boolean(object.isPreview) : false,
    };
  }

  export function toJSON(message: RealmInfo): unknown {
    const obj: any = {};
    message.baseUrl !== undefined && (obj.baseUrl = message.baseUrl);
    message.realmName !== undefined && (obj.realmName = message.realmName);
    message.networkId !== undefined && (obj.networkId = Math.round(message.networkId));
    message.commsAdapter !== undefined && (obj.commsAdapter = message.commsAdapter);
    message.isPreview !== undefined && (obj.isPreview = message.isPreview);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<RealmInfo>, I>>(base?: I): RealmInfo {
    return RealmInfo.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<RealmInfo>, I>>(object: I): RealmInfo {
    const message = createBaseRealmInfo();
    message.baseUrl = object.baseUrl ?? "";
    message.realmName = object.realmName ?? "";
    message.networkId = object.networkId ?? 0;
    message.commsAdapter = object.commsAdapter ?? "";
    message.isPreview = object.isPreview ?? false;
    return message;
  }
}

function createBaseGetRealmResponse(): GetRealmResponse {
  return { realmInfo: undefined };
}

export namespace GetRealmResponse {
  export function encode(message: GetRealmResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.realmInfo !== undefined) {
      RealmInfo.encode(message.realmInfo, writer.uint32(10).fork()).ldelim();
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): GetRealmResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGetRealmResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.realmInfo = RealmInfo.decode(reader, reader.uint32());
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): GetRealmResponse {
    return { realmInfo: isSet(object.realmInfo) ? RealmInfo.fromJSON(object.realmInfo) : undefined };
  }

  export function toJSON(message: GetRealmResponse): unknown {
    const obj: any = {};
    message.realmInfo !== undefined &&
      (obj.realmInfo = message.realmInfo ? RealmInfo.toJSON(message.realmInfo) : undefined);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<GetRealmResponse>, I>>(base?: I): GetRealmResponse {
    return GetRealmResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<GetRealmResponse>, I>>(object: I): GetRealmResponse {
    const message = createBaseGetRealmResponse();
    message.realmInfo = (object.realmInfo !== undefined && object.realmInfo !== null)
      ? RealmInfo.fromPartial(object.realmInfo)
      : undefined;
    return message;
  }
}

function createBaseGetWorldTimeResponse(): GetWorldTimeResponse {
  return { seconds: 0 };
}

export namespace GetWorldTimeResponse {
  export function encode(message: GetWorldTimeResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.seconds !== 0) {
      writer.uint32(8).int32(message.seconds);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): GetWorldTimeResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGetWorldTimeResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 8) {
            break;
          }

          message.seconds = reader.int32();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): GetWorldTimeResponse {
    return { seconds: isSet(object.seconds) ? Number(object.seconds) : 0 };
  }

  export function toJSON(message: GetWorldTimeResponse): unknown {
    const obj: any = {};
    message.seconds !== undefined && (obj.seconds = Math.round(message.seconds));
    return obj;
  }

  export function create<I extends Exact<DeepPartial<GetWorldTimeResponse>, I>>(base?: I): GetWorldTimeResponse {
    return GetWorldTimeResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<GetWorldTimeResponse>, I>>(object: I): GetWorldTimeResponse {
    const message = createBaseGetWorldTimeResponse();
    message.seconds = object.seconds ?? 0;
    return message;
  }
}

function createBaseGetRealmRequest(): GetRealmRequest {
  return {};
}

export namespace GetRealmRequest {
  export function encode(_: GetRealmRequest, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): GetRealmRequest {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGetRealmRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(_: any): GetRealmRequest {
    return {};
  }

  export function toJSON(_: GetRealmRequest): unknown {
    const obj: any = {};
    return obj;
  }

  export function create<I extends Exact<DeepPartial<GetRealmRequest>, I>>(base?: I): GetRealmRequest {
    return GetRealmRequest.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<GetRealmRequest>, I>>(_: I): GetRealmRequest {
    const message = createBaseGetRealmRequest();
    return message;
  }
}

function createBaseGetWorldTimeRequest(): GetWorldTimeRequest {
  return {};
}

export namespace GetWorldTimeRequest {
  export function encode(_: GetWorldTimeRequest, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): GetWorldTimeRequest {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseGetWorldTimeRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(_: any): GetWorldTimeRequest {
    return {};
  }

  export function toJSON(_: GetWorldTimeRequest): unknown {
    const obj: any = {};
    return obj;
  }

  export function create<I extends Exact<DeepPartial<GetWorldTimeRequest>, I>>(base?: I): GetWorldTimeRequest {
    return GetWorldTimeRequest.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<GetWorldTimeRequest>, I>>(_: I): GetWorldTimeRequest {
    const message = createBaseGetWorldTimeRequest();
    return message;
  }
}

function createBaseReadFileRequest(): ReadFileRequest {
  return { fileName: "" };
}

export namespace ReadFileRequest {
  export function encode(message: ReadFileRequest, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.fileName !== "") {
      writer.uint32(10).string(message.fileName);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): ReadFileRequest {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseReadFileRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.fileName = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): ReadFileRequest {
    return { fileName: isSet(object.fileName) ? String(object.fileName) : "" };
  }

  export function toJSON(message: ReadFileRequest): unknown {
    const obj: any = {};
    message.fileName !== undefined && (obj.fileName = message.fileName);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<ReadFileRequest>, I>>(base?: I): ReadFileRequest {
    return ReadFileRequest.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<ReadFileRequest>, I>>(object: I): ReadFileRequest {
    const message = createBaseReadFileRequest();
    message.fileName = object.fileName ?? "";
    return message;
  }
}

function createBaseReadFileResponse(): ReadFileResponse {
  return { content: new Uint8Array(), hash: "" };
}

export namespace ReadFileResponse {
  export function encode(message: ReadFileResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.content.length !== 0) {
      writer.uint32(10).bytes(message.content);
    }
    if (message.hash !== "") {
      writer.uint32(18).string(message.hash);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): ReadFileResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseReadFileResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.content = reader.bytes();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.hash = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): ReadFileResponse {
    return {
      content: isSet(object.content) ? bytesFromBase64(object.content) : new Uint8Array(),
      hash: isSet(object.hash) ? String(object.hash) : "",
    };
  }

  export function toJSON(message: ReadFileResponse): unknown {
    const obj: any = {};
    message.content !== undefined &&
      (obj.content = base64FromBytes(message.content !== undefined ? message.content : new Uint8Array()));
    message.hash !== undefined && (obj.hash = message.hash);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<ReadFileResponse>, I>>(base?: I): ReadFileResponse {
    return ReadFileResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<ReadFileResponse>, I>>(object: I): ReadFileResponse {
    const message = createBaseReadFileResponse();
    message.content = object.content ?? new Uint8Array();
    message.hash = object.hash ?? "";
    return message;
  }
}

function createBaseCurrentSceneEntityRequest(): CurrentSceneEntityRequest {
  return {};
}

export namespace CurrentSceneEntityRequest {
  export function encode(_: CurrentSceneEntityRequest, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): CurrentSceneEntityRequest {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseCurrentSceneEntityRequest();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(_: any): CurrentSceneEntityRequest {
    return {};
  }

  export function toJSON(_: CurrentSceneEntityRequest): unknown {
    const obj: any = {};
    return obj;
  }

  export function create<I extends Exact<DeepPartial<CurrentSceneEntityRequest>, I>>(
    base?: I,
  ): CurrentSceneEntityRequest {
    return CurrentSceneEntityRequest.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<CurrentSceneEntityRequest>, I>>(
    _: I,
  ): CurrentSceneEntityRequest {
    const message = createBaseCurrentSceneEntityRequest();
    return message;
  }
}

function createBaseCurrentSceneEntityResponse(): CurrentSceneEntityResponse {
  return { urn: "", content: [], metadataJson: "", baseUrl: "" };
}

export namespace CurrentSceneEntityResponse {
  export function encode(message: CurrentSceneEntityResponse, writer: _m0.Writer = _m0.Writer.create()): _m0.Writer {
    if (message.urn !== "") {
      writer.uint32(10).string(message.urn);
    }
    for (const v of message.content) {
      ContentMapping.encode(v!, writer.uint32(18).fork()).ldelim();
    }
    if (message.metadataJson !== "") {
      writer.uint32(26).string(message.metadataJson);
    }
    if (message.baseUrl !== "") {
      writer.uint32(34).string(message.baseUrl);
    }
    return writer;
  }

  export function decode(input: _m0.Reader | Uint8Array, length?: number): CurrentSceneEntityResponse {
    const reader = input instanceof _m0.Reader ? input : _m0.Reader.create(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseCurrentSceneEntityResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          if (tag !== 10) {
            break;
          }

          message.urn = reader.string();
          continue;
        case 2:
          if (tag !== 18) {
            break;
          }

          message.content.push(ContentMapping.decode(reader, reader.uint32()));
          continue;
        case 3:
          if (tag !== 26) {
            break;
          }

          message.metadataJson = reader.string();
          continue;
        case 4:
          if (tag !== 34) {
            break;
          }

          message.baseUrl = reader.string();
          continue;
      }
      if ((tag & 7) === 4 || tag === 0) {
        break;
      }
      reader.skipType(tag & 7);
    }
    return message;
  }

  export function fromJSON(object: any): CurrentSceneEntityResponse {
    return {
      urn: isSet(object.urn) ? String(object.urn) : "",
      content: Array.isArray(object?.content) ? object.content.map((e: any) => ContentMapping.fromJSON(e)) : [],
      metadataJson: isSet(object.metadataJson) ? String(object.metadataJson) : "",
      baseUrl: isSet(object.baseUrl) ? String(object.baseUrl) : "",
    };
  }

  export function toJSON(message: CurrentSceneEntityResponse): unknown {
    const obj: any = {};
    message.urn !== undefined && (obj.urn = message.urn);
    if (message.content) {
      obj.content = message.content.map((e) => e ? ContentMapping.toJSON(e) : undefined);
    } else {
      obj.content = [];
    }
    message.metadataJson !== undefined && (obj.metadataJson = message.metadataJson);
    message.baseUrl !== undefined && (obj.baseUrl = message.baseUrl);
    return obj;
  }

  export function create<I extends Exact<DeepPartial<CurrentSceneEntityResponse>, I>>(
    base?: I,
  ): CurrentSceneEntityResponse {
    return CurrentSceneEntityResponse.fromPartial(base ?? {});
  }

  export function fromPartial<I extends Exact<DeepPartial<CurrentSceneEntityResponse>, I>>(
    object: I,
  ): CurrentSceneEntityResponse {
    const message = createBaseCurrentSceneEntityResponse();
    message.urn = object.urn ?? "";
    message.content = object.content?.map((e) => ContentMapping.fromPartial(e)) || [];
    message.metadataJson = object.metadataJson ?? "";
    message.baseUrl = object.baseUrl ?? "";
    return message;
  }
}

export type RuntimeServiceDefinition = typeof RuntimeServiceDefinition;
export const RuntimeServiceDefinition = {
  name: "RuntimeService",
  fullName: "decentraland.kernel.apis.RuntimeService",
  methods: {
    /** Provides information about the current realm */
    getRealm: {
      name: "GetRealm",
      requestType: GetRealmRequest,
      requestStream: false,
      responseType: GetRealmResponse,
      responseStream: false,
      options: {},
    },
    /**
     * Provides information about the Decentraland Time, which is coordinated
     * across players.
     */
    getWorldTime: {
      name: "GetWorldTime",
      requestType: GetWorldTimeRequest,
      requestStream: false,
      responseType: GetWorldTimeResponse,
      responseStream: false,
      options: {},
    },
    /**
     * Returns the file content of a deployed asset. If the file doesn't
     * exist or cannot be retrieved, the RPC call throws an error.
     * This method is called to load any assets deployed among the scene,
     * runtime may cache this response much more than the provided "fetch" function.
     */
    readFile: {
      name: "ReadFile",
      requestType: ReadFileRequest,
      requestStream: false,
      responseType: ReadFileResponse,
      responseStream: false,
      options: {},
    },
    /** Returns information about the current scene. This is the replacement of GetBootstrapData */
    getSceneInformation: {
      name: "GetSceneInformation",
      requestType: CurrentSceneEntityRequest,
      requestStream: false,
      responseType: CurrentSceneEntityResponse,
      responseStream: false,
      options: {},
    },
  },
} as const;

declare var self: any | undefined;
declare var window: any | undefined;
declare var global: any | undefined;
var tsProtoGlobalThis: any = (() => {
  if (typeof globalThis !== "undefined") {
    return globalThis;
  }
  if (typeof self !== "undefined") {
    return self;
  }
  if (typeof window !== "undefined") {
    return window;
  }
  if (typeof global !== "undefined") {
    return global;
  }
  throw "Unable to locate global object";
})();

function bytesFromBase64(b64: string): Uint8Array {
  if (tsProtoGlobalThis.Buffer) {
    return Uint8Array.from(tsProtoGlobalThis.Buffer.from(b64, "base64"));
  } else {
    const bin = tsProtoGlobalThis.atob(b64);
    const arr = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; ++i) {
      arr[i] = bin.charCodeAt(i);
    }
    return arr;
  }
}

function base64FromBytes(arr: Uint8Array): string {
  if (tsProtoGlobalThis.Buffer) {
    return tsProtoGlobalThis.Buffer.from(arr).toString("base64");
  } else {
    const bin: string[] = [];
    arr.forEach((byte) => {
      bin.push(String.fromCharCode(byte));
    });
    return tsProtoGlobalThis.btoa(bin.join(""));
  }
}

type Builtin = Date | Function | Uint8Array | string | number | boolean | undefined;

export type DeepPartial<T> = T extends Builtin ? T
  : T extends Array<infer U> ? Array<DeepPartial<U>> : T extends ReadonlyArray<infer U> ? ReadonlyArray<DeepPartial<U>>
  : T extends { $case: string } ? { [K in keyof Omit<T, "$case">]?: DeepPartial<T[K]> } & { $case: T["$case"] }
  : T extends {} ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & { [K in Exclude<keyof I, KeysOfUnion<P>>]: never };

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
